olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -662,6 +663,68 @@ VersionedLayerClientImpl::GetAggregatedData(TileRequest request) {
   return client::CancellableFuture<AggregatedDataResponse>(
       std::move(cancel_token), std::move(promise));
 }
+
+client::CancellationToken VersionedLayerClientImpl::Protect(
+    const TileKeys& tiles, ProtectCallback callback) {
+  auto version = catalog_version_.load();
+  auto protect_task =
+      [=](client::CancellationContext context) -> ProtectResponse {
+    // could protect keys, which is already in cache
+    // because we don't know quad tree which will contain tiles
+    repository::DataCacheRepository data_cache_repository(catalog_,
+                                                          settings_.cache);
+    repository::PartitionsCacheRepository partitions_cache_repository(
+        catalog_, settings_.cache);
+
+    cache::KeyValueCache::KeyListType keys_to_protect;
+    keys_to_protect.reserve(tiles.size() * 2);
+
+    std::vector<read::QuadTreeIndex> quad_trees;
+    quad_trees.reserve(tiles.size());
+    for (const auto& tile : tiles) {
+      auto it = std::find_if(
+          quad_trees.begin(), quad_trees.end(),
+          [&tile](const read::QuadTreeIndex& quad) {
+            auto root = quad.GetRootTile();
+            return (tile.Level() >= root.Level() &&
+                    tile.Level() - root.Level() <= kQuadTreeDepth &&
+                    root.IsParentOf(tile));
```

Comment:
How about extending the QuadTreeIndex with a `bool Contains(TileKey)` method instead of "assuming" the tile is included. 
If it is enough to just look if the quadtree root is your parent, then I would instead just look if the quadtree is cached and work with string keys instead of QuadTreeIndex in your `quad_trees` vector. And then each time you have a new tile just check if one of its parent's key is already present in `quad_trees`.

Reply:
This what is actually checked if root lile in quad tree is parent. Should I move it to QuadTreeIndex, that it could be used in other cases?

Reply:
I dont think it is enough to just check if it is parent as there are layers without certain data, e.g. only level 12 tiles, or only level 14 tiles, etc. If you check there that a level 14 tile is a child of a level 10 root but you will never have these tiles downloaded, then what is the point? You should check that the tile to be protected is actually inside the quad tree index and for this i would propose to add a new method in QuadTreeIndex.